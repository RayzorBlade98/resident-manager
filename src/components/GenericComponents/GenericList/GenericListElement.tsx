import { StyleAttribute, merge } from 'glamor';
import React, { PropsWithChildren } from 'react';
import styles from './styles';

interface GenericListElementProps {
  /**
   * onClick callback
   */
  onClick: () => void;

  /**
   * Whether the element is selected or not
   */
  selected?: boolean;

  /**
   * Additional stlye of the list element
   */
  style?: StyleAttribute;
}

/**
 * Component that displays an element inside of a `GenericList`
 */
function GenericListElement(
  props: PropsWithChildren<GenericListElementProps>,
): JSX.Element {
  return (
    <div
      onClick={props.onClick}
      {...merge(
        props.selected ? styles.selectedListElement : styles.listElement,
        props.style,
      )}
      className={`listElement${props.selected ? ' selectedListElement' : ''}`}
    >
      {props.children}
    </div>
  );
}

export default GenericListElement;
