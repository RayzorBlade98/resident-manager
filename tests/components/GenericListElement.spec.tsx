/* eslint-disable max-len */

import {
  RenderResult, render, prettyDOM, fireEvent,
} from '@testing-library/react';
import { css } from 'glamor';
import React from 'react';
import GenericListElement from '_/components/GenericComponents/GenericList/GenericListElement';

describe('GenericListElement', () => {
  let renderResult: RenderResult;
  let onClickMock: jest.Mock;
  let elementContent: JSX.Element;

  function setupRendering(selected = false): void {
    onClickMock = jest.fn();
    elementContent = <div>Element Content</div>;
    renderResult = render(
      <GenericListElement
        style={css({ color: 'red' })}
        onClick={onClickMock}
        selected={selected}
      >
        {elementContent}
      </GenericListElement>,
    );
  }

  beforeEach(() => {
    setupRendering();
  });

  test('should render children correctly', () => {
    // Arrange
    const expectedElementContent = render(elementContent).container.firstElementChild!;

    // Assert
    const actualElementContent = renderResult.container.firstElementChild!.firstElementChild!;
    expect(prettyDOM(actualElementContent)).toEqual(
      prettyDOM(expectedElementContent),
    );
  });

  test('should call onClick callback', () => {
    // Act
    fireEvent.click(renderResult.container.firstElementChild!);

    // Assert
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  test('should match snapshot', () => {
    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });

  test('should match snapshot (selected)', () => {
    // Arrange
    setupRendering(true);

    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });
});
