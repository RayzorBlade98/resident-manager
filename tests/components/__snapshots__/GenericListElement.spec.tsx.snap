// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericListElement should match snapshot (selected) 1`] = `
.css-0,
[data-css-0] {
  width: 100%;
  height: 10%;
  color: red;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #c8daea;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.css-0:hover,
[data-css-0]:hover {
  background-color: #c8daea;
  cursor: pointer;
}

<div>
  <div
    class="listElement selectedListElement"
    data-css-0=""
  >
    <div>
      Element Content
    </div>
  </div>
</div>
`;

exports[`GenericListElement should match snapshot 1`] = `
.css-0,
[data-css-0] {
  width: 100%;
  height: 10%;
  color: red;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.css-0:hover,
[data-css-0]:hover {
  background-color: #c8daea;
  cursor: pointer;
}

<div>
  <div
    class="listElement"
    data-css-0=""
  >
    <div>
      Element Content
    </div>
  </div>
</div>
`;
